From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Oleg <balbes-150@yandex.ru>
Date: Thu, 7 Oct 2021 16:48:09 +0200
Subject: [ARCHEOLOGY] fix analog sound station-p1 (#3166)

> X-Git-Archeology: - Revision b2d54059f75ae69b71a6a422c178ceef271336f1: https://github.com/armbian/build/commit/b2d54059f75ae69b71a6a422c178ceef271336f1
> X-Git-Archeology:   Date: Thu, 07 Oct 2021 16:48:09 +0200
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: fix analog sound station-p1 (#3166)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 6b490e16944b30ff69bf9c13678905187df0d9d4: https://github.com/armbian/build/commit/6b490e16944b30ff69bf9c13678905187df0d9d4
> X-Git-Archeology:   Date: Tue, 11 Jan 2022 15:26:11 +0100
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: move kernel edge to 5.16 (#3387)
> X-Git-Archeology:
> X-Git-Archeology: - Revision f52a4193d02ef88333ba117c68d49486dfd7ff41: https://github.com/armbian/build/commit/f52a4193d02ef88333ba117c68d49486dfd7ff41
> X-Git-Archeology:   Date: Sun, 20 Mar 2022 22:58:21 +0100
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: Adding Pine64 Quartz64a as WIP target (#3539)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 0afe24c95729044910e0b3f84dc5500bcdc6524c: https://github.com/armbian/build/commit/0afe24c95729044910e0b3f84dc5500bcdc6524c
> X-Git-Archeology:   Date: Sun, 24 Apr 2022 22:33:47 +0200
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: move kernel media edge to 5.17 (#3704)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 2e1b4eed739d6ea81c3a7ba7e94d1bcab5cbbb6c: https://github.com/armbian/build/commit/2e1b4eed739d6ea81c3a7ba7e94d1bcab5cbbb6c
> X-Git-Archeology:   Date: Mon, 04 Jul 2022 16:18:31 +0300
> X-Git-Archeology:   From: balbes150 <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: move kernel media to current 5.18 and edge 5.19
> X-Git-Archeology:
> X-Git-Archeology: - Revision 738c1613073bfd504aebc223a0d110cc6afefbe5: https://github.com/armbian/build/commit/738c1613073bfd504aebc223a0d110cc6afefbe5
> X-Git-Archeology:   Date: Sat, 12 Nov 2022 10:11:22 +0100
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: move kernel media-current to 5.19.17 (#4406)
> X-Git-Archeology:
> X-Git-Archeology: - Revision 276f346ee9c21cdcd627e54676334f4d5a9292b6: https://github.com/armbian/build/commit/276f346ee9c21cdcd627e54676334f4d5a9292b6
> X-Git-Archeology:   Date: Wed, 14 Dec 2022 10:17:09 +0100
> X-Git-Archeology:   From: Oleg <balbes-150@yandex.ru>
> X-Git-Archeology:   Subject: move kernel media to current 6.0 and edge 6.1 (#4566)
> X-Git-Archeology:
---
 sound/soc/codecs/es8328.c | 114 ++++++++++
 1 file changed, 114 insertions(+)

diff --git a/sound/soc/codecs/es8328.c b/sound/soc/codecs/es8328.c
index 160adc706cc6..c0dbda696e9b 100644
--- a/sound/soc/codecs/es8328.c
+++ b/sound/soc/codecs/es8328.c
@@ -21,8 +21,13 @@
 #include <sound/pcm_params.h>
 #include <sound/soc.h>
 #include <sound/tlv.h>
+#include <linux/of_gpio.h>
+#include <linux/gpio.h>
 #include "es8328.h"
 
+#define INVALID_GPIO		   -1
+#define ES8328_CODEC_SET_HP		1
+
 static const unsigned int rates_12288[] = {
 	8000, 12000, 16000, 24000, 32000, 48000, 96000,
 };
@@ -86,8 +91,18 @@ struct es8328_priv {
 	const int *mclk_ratios;
 	bool provider;
 	struct regulator_bulk_data supplies[ES8328_SUPPLY_NUM];
+
+    int hp_ctl_gpio;
+    int hp_det_gpio;
+
+    bool muted;
+    bool hp_inserted;
+    bool hp_gpio_level;
+    bool hp_det_level;
 };
 
+static struct es8328_priv *es8328_private;
+
 /*
  * ES8328 Controls
  */
@@ -112,6 +127,42 @@ static const struct {
 	{ 48000, ES8328_DACCONTROL6_DEEMPH_48k },
 };
 
+static int es8328_set_gpio(int gpio, bool level)
+{
+	struct es8328_priv *es8328 = es8328_private;
+
+	if (!es8328) {
+		return 0;
+	}
+
+	if ((gpio & ES8328_CODEC_SET_HP) && es8328
+	    && es8328->hp_ctl_gpio != INVALID_GPIO) {
+		gpio_set_value(es8328->hp_ctl_gpio, level);
+	}
+
+	return 0;
+}
+
+static irqreturn_t hp_det_irq_handler(int irq, void *dev_id)
+{
+	struct es8328_priv *es8328 = es8328_private;
+
+    if(gpio_get_value(es8328->hp_det_gpio)) {
+        es8328->hp_inserted = 1;
+    } else {
+        es8328->hp_inserted = 0;
+    }
+
+	if(!es8328->muted && es8328->hp_inserted) {
+		es8328_set_gpio(ES8328_CODEC_SET_HP, es8328->hp_gpio_level);
+	} else {
+        es8328_set_gpio(ES8328_CODEC_SET_HP, !es8328->hp_gpio_level);
+    }
+	return IRQ_HANDLED;
+}
+
+
+
 static int es8328_set_deemph(struct snd_soc_component *component)
 {
 	struct es8328_priv *es8328 = snd_soc_component_get_drvdata(component);
@@ -454,6 +505,14 @@ static const struct snd_soc_dapm_route es8328_dapm_routes[] = {
 
 static int es8328_mute(struct snd_soc_dai *dai, int mute, int direction)
 {
+    struct es8328_priv *es8328 = snd_soc_component_get_drvdata(dai->component);
+    es8328->muted = mute;
+    if (!mute && es8328->hp_inserted) {
+        es8328_set_gpio(ES8328_CODEC_SET_HP, es8328->hp_gpio_level);
+    } else {
+        es8328_set_gpio(ES8328_CODEC_SET_HP, !es8328->hp_gpio_level);
+    }
+
 	return snd_soc_component_update_bits(dai->component, ES8328_DACCONTROL3,
 			ES8328_DACCONTROL3_DACMUTE,
 			mute ? ES8328_DACCONTROL3_DACMUTE : 0);
@@ -798,6 +857,21 @@ static int es8328_component_probe(struct snd_soc_component *component)
 		goto clk_fail;
 	}
 
+    if (es8328->hp_det_gpio != INVALID_GPIO) {
+		if (gpio_get_value(es8328->hp_det_gpio) == es8328->hp_det_level)
+			es8328->hp_inserted = 1;
+	} else {
+        es8328->hp_inserted = 1;
+    }
+        
+ 
+    if (!strncmp(component->dev->of_node->name, "es8388", 6)) {
+        usleep_range(18000, 20000);
+        snd_soc_component_update_bits(component, ES8328_DACCONTROL17,
+			    ES8328_DACCONTROL17_LD2LO, ES8328_DACCONTROL17_LD2LO);
+        snd_soc_component_update_bits(component, ES8328_DACCONTROL20,
+			    ES8328_DACCONTROL20_RD2RO, ES8328_DACCONTROL20_RD2RO);
+    }
 	return 0;
 
 clk_fail:
@@ -851,6 +925,8 @@ int es8328_probe(struct device *dev, struct regmap *regmap)
 	struct es8328_priv *es8328;
 	int ret;
 	int i;
+    int hp_irq = 0;
+    enum of_gpio_flags flags;
 
 	if (IS_ERR(regmap))
 		return PTR_ERR(regmap);
@@ -860,6 +936,7 @@ int es8328_probe(struct device *dev, struct regmap *regmap)
 		return -ENOMEM;
 
 	es8328->regmap = regmap;
+    es8328_private = es8328;
 
 	for (i = 0; i < ARRAY_SIZE(es8328->supplies); i++)
 		es8328->supplies[i].supply = supply_names[i];
@@ -871,6 +948,43 @@ int es8328_probe(struct device *dev, struct regmap *regmap)
 		return ret;
 	}
 
+    es8328->hp_ctl_gpio = of_get_named_gpio_flags(dev->of_node, "hp-ctl-gpio", 0, &flags);
+	if (es8328->hp_ctl_gpio < 0) {
+		dev_info(dev, "Can not read property hp_ctl_gpio\n");
+		es8328->hp_ctl_gpio = INVALID_GPIO;
+	} else {
+		es8328->hp_gpio_level = (flags & OF_GPIO_ACTIVE_LOW) ? 0 : 1;
+		ret = devm_gpio_request_one(dev, es8328->hp_ctl_gpio, GPIOF_DIR_OUT, "hp_ctl_gpio");
+		if (ret != 0) {
+			dev_err(dev, "Failed to request hp_ctl_gpio\n");
+			return ret;
+		}
+		es8328_set_gpio(ES8328_CODEC_SET_HP, !es8328->hp_gpio_level);
+	}
+
+	es8328->hp_det_gpio = of_get_named_gpio_flags(dev->of_node, "hp-det-gpio", 0, &flags);
+	if (es8328->hp_det_gpio < 0) {
+		dev_info(dev, "Can not read property hp_det_gpio\n");
+		es8328->hp_det_gpio = INVALID_GPIO;
+	} else {
+		es8328->hp_det_level = (flags & OF_GPIO_ACTIVE_LOW) ? 0 : 1;
+		ret = devm_gpio_request_one(dev, es8328->hp_det_gpio, GPIOF_IN, NULL);
+		if (ret != 0) {
+			dev_err(dev, "Failed to request hp_det_gpio\n");
+			return ret;
+		}
+		hp_irq = gpio_to_irq(es8328->hp_det_gpio);
+
+		if (hp_irq) {
+			ret = devm_request_threaded_irq(dev, hp_irq, NULL, hp_det_irq_handler,
+					IRQ_TYPE_EDGE_BOTH | IRQF_ONESHOT, "ES8323", NULL);
+			if (ret < 0) {
+				dev_err(dev, "request_irq failed: %d\n", ret);
+				return ret;
+			}
+		}
+	}
+
 	dev_set_drvdata(dev, es8328);
 
 	return devm_snd_soc_register_component(dev,
-- 
Armbian

